### ISSUE TITLE
Arch v3 Docker Compose Nomenclature: dynamic, use of `COMPOSE_PROJECT_NAME` and no `container_name`

### ISSUE THREAD

**Issue by mrjones-plip** [medic/cht-core#7802]
**Describe the issue**
When reviewing the new [cht-core](https://staging.dev.medicmobile.org/_couch/builds/medic%3Amedic%3Aarchv3/docker-compose%2Fcht-core.yml) and [cht-couchdb](https://staging.dev.medicmobile.org/_couch/builds/medic%3Amedic%3Aarchv3/docker-compose%2Fcht-couchdb.yml) docker-compose files for archv3 currently published to staging.dev.medicmobile.org, I'd prefer not to use an explicit name for the networks and the volumes. These can be [dynamically set within a project](https://www.google.com/url?q=https://docs.docker.com/engine/reference/commandline/compose/%23use--p-to-specify-a-project-name&sa=D&source=calendar&ust=1663791886590864&usg=AOvVaw2VGhCwB747h1fvOXGxdOaX) when using `COMPOSE_PROJECT_NAME` but only if they're not explicitly set in the compose file. This is a huge win for both multi-tenant and developer setups (which are often multi-tenant ; ) 

IE go from:
```yaml
   networks: 
      cht-net: 
         name: cht-net 
```

 to this 

```yaml
   networks: 
      cht-net: 
```

Which ever way we go with, declaring networks and volumes they should be uniform.  The CouchDB container and the cht-core containers differ. (note extra - and lacking :).  My vote is the former!

CouchDB:

```yaml
   networks: 
      cht-net: 
```

CHT Core:

```yaml
   networks: 
      - cht-net
```

While we consider using `COMPOSE_PROJECT_NAME`, I recommend _not_ using `container_name` for each container.  They will already get named after the outer most stanza value (the service) and in every case the `container_name` matches the service (eg: `container_name: cht-nginx` == `cht-nginx:` ).  Intra-container lookups will [not be affected](https://www.google.com/url?q=https://stackoverflow.com/questions/55522620/docker-compose-yml-container-name-and-hostname&sa=D&source=calendar&ust=1663791886590864&usg=AOvVaw1QRbFqBX0Cf4GwSEcseprF).

Further, implicit nomenclature will simplify scenarios where we have service name <-> container name mismatches.  For example, `haproxy` service defaults to using the docker DNS of `couchdb `as seen here - `"COUCHDB_SERVERS=${COUCHDB_SERVERS:-couchdb}` but the container is actually called something different (`container_name: cht-couchdb`).  The reason this is currently working is that DNS works with both service name and container name.

**Describe alternatives you've considered**
There may have been intentional design decisions that I don't appreciate that mandate sticking with the current nomenclature

**Comment by garethbowen** [comment]
IMO this shouldn't block the 4.0.0 release - marking as low priority.

**Comment by dianabarsan** [comment]
This _was_ a design decision to make sure docker-compose mapped containers correctly even when the containers were started by the cht-upgrade-service, but I've tested it today and it seems to work also when using project names. 

I think it's a fairly simple change, and worth the exploration. 
Since this affects both CHT-Core and cht-upgrade-service, I think it's best if we decide on the direction now, rather than later and need everyone upgrade their upgrade-service as well.

**Comment by mrjones-plip** [comment]
Stumbled upon [this good quote](https://stackoverflow.com/a/57361318) (emphasis mine):

> Docker Compose normally assigns a container name based on its current project name and the name of the services: block. Specifying container_name: explicitly overrides this; but, it means you can’t launch multiple copies of the same Compose file with different project names (from different directories) because the container name you’ve explicitly chosen won’t be used.
> 
> **You almost never care what the container name is explicitly. It only really matters if you’re trying to use plain docker commands to manipulate Compose-managed containers; it has no impact on inter-service communication. Just delete the container_name: line.**
> 
> (For similar reasons you can almost always delete hostname: and links: sections if you have them with no practical impact on your overall system.)

**Comment by dianabarsan** [comment]
Unfortunately, I don't think we can't use the dynamically generated docker network. That's because, in self hosted environments, we have to run the `cht-upgrade-service` separately (a separate docker-compose file in a different folder, which we run separately: https://github.com/medic/cht-upgrade-service/blob/main/docker-compose.yml). For security purposes, we don't expose the upgrade service port to the host, so it would only be accessible from within the network (since we can't authenticate these requests). 
Is there a way that the cht-upgrade-service can still remain isolated from the host and accessible within the "cht" network?

**Comment by dianabarsan** [comment]
I'm going to try to use an environment variable to define the network first.

**Comment by mrjones-plip** [comment]
> Is there a way that the cht-upgrade-service can still remain isolated from the host and accessible within the "cht" network?

I think you hopefully answered your own question with:

> I'm going to try to use an environment variable to define the network first.

But let me know!

**Comment by mrjones-plip** [comment]
> Is there a way that the cht-upgrade-service can still remain isolated from the host and accessible within the "cht" network?

tl;dr - if you start the upgrade service with the same `COMPOSE_PROJECT_NAME` as you did with your other containers,  everything  _should_ just work. Mainly, services can be resolved by the outer most name (eg `cht-nginx:`) and no explicit name is needed (`container_name: cht-nginx`) as long as they're on the same network. Unauthenticated calls using the default service name (with out project name)  should also work.  Further, other compose instances on different networks (as defined by different `COMPOSE_PROJECT_NAME` values) will not be able to access other docker networks.

Background on this - I was testing when I was filing out #7832 and discovered that with a modified compose file and an env var of  `COMPOSE_PROJECT_NAME=testee` you get a volumes of `testee_cht-ssl` and `testee_cht-credentials`,  network of `testee_cht-net` and then containers to match (per `docker ps --format '{{.Names}}'|grep test`):

```                                 
testee_couchdb_1
testee_cht-nginx_1
testee_cht-sentinel_1
testee_cht-api_1
testee_haproxy_1
testee_healthcheck_1
```

To fully reproduce my test environment:

<details>
<summary> grab this core compose file </summary>

```yaml
version: '3.9'

services:
  haproxy:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-haproxy:4.0.0-alpha
    environment:
      - "HAPROXY_IP=${HAPROXY_IP:-haproxy}"
      - "COUCHDB_USER=${COUCHDB_USER:-admin}"
      - "COUCHDB_PASSWORD=${COUCHDB_PASSWORD}"
      - "COUCHDB_SERVERS=${COUCHDB_SERVERS:-couchdb}"
      - "HAPROXY_PORT=${HAPROXY_PORT:-5984}"
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:
    expose:
      - ${HAPROXY_PORT:-5984}

  healthcheck:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-haproxy-healthcheck:4.0.0-alpha
    environment:
      - "COUCHDB_SERVERS=${COUCHDB_SERVERS:-couchdb}"
      - "COUCHDB_USER=${COUCHDB_USER:-admin}"
      - "COUCHDB_PASSWORD=${COUCHDB_PASSWORD}"
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:

  cht-api:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-api:4.0.0-alpha
    depends_on:
      - haproxy
    expose:
      - "${API_PORT:-5988}"
    environment:
      - COUCH_URL=http://${COUCHDB_USER:-admin}:${COUCHDB_PASSWORD}@haproxy:${HAPROXY_PORT:-5984}/medic
      - BUILDS_URL=${MARKET_URL_READ:-https://staging.dev.medicmobile.org}/${BUILDS_SERVER:-_couch/builds}
      - UPGRADE_SERVICE_URL=${UPGRADE_SERVICE_URL:-http://localhost:5100}
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:

  cht-sentinel:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-sentinel:4.0.0-alpha
    depends_on:
      - haproxy
    environment:
      - COUCH_URL=http://${COUCHDB_USER:-admin}:${COUCHDB_PASSWORD}@haproxy:${HAPROXY_PORT:-5984}/medic
      - API_HOST=cht-api
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:

  cht-nginx:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-nginx:4.0.0-alpha
    depends_on:
      - cht-api
      - haproxy
    ports:
      - "${NGINX_HTTP_PORT:-80}:80"
      - "${NGINX_HTTPS_PORT:-443}:443"
    volumes:
      - cht-ssl:/etc/nginx/private/
    environment:
      - "CERTIFICATE_MODE=${CERTIFICATE_MODE:-SELF_SIGNED}"
      - "SSL_CERT_FILE_PATH=${SSL_CERT_FILE_PATH:-/etc/nginx/private/cert.pem}"
      - "SSL_KEY_FILE_PATH=${SSL_KEY_FILE_PATH:-/etc/nginx/private/key.pem}"
      - "COMMON_NAME=${COMMON_NAME:-test-nginx.dev.medicmobile.org}"
      - "EMAIL=${EMAIL:-domains@medic.org}"
      - "COUNTRY=${COUNTRY:-US}"
      - "STATE=${STATE:-California}"
      - "LOCALITY=${LOCALITY:-San_Francisco}"
      - "ORGANISATION=${ORGANISATION:-medic}"
      - "DEPARTMENT=${DEPARTMENT:-Information_Security}"
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:

networks:
  cht-net:

volumes:
  cht-ssl:
```

</details> 


<details>
<summary> grab this couchdb compose file </summary>

```yaml
version: '3.9'

services:
  haproxy:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-haproxy:4.0.0-alpha
    environment:
      - "HAPROXY_IP=${HAPROXY_IP:-haproxy}"
      - "COUCHDB_USER=${COUCHDB_USER:-admin}"
      - "COUCHDB_PASSWORD=${COUCHDB_PASSWORD}"
      - "COUCHDB_SERVERS=${COUCHDB_SERVERS:-couchdb}"
      - "HAPROXY_PORT=${HAPROXY_PORT:-5984}"
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:
    expose:
      - ${HAPROXY_PORT:-5984}

  healthcheck:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-haproxy-healthcheck:4.0.0-alpha
    environment:
      - "COUCHDB_SERVERS=${COUCHDB_SERVERS:-couchdb}"
      - "COUCHDB_USER=${COUCHDB_USER:-admin}"
      - "COUCHDB_PASSWORD=${COUCHDB_PASSWORD}"
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:

  cht-api:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-api:4.0.0-alpha
    depends_on:
      - haproxy
    expose:
      - "${API_PORT:-5988}"
    environment:
      - COUCH_URL=http://${COUCHDB_USER:-admin}:${COUCHDB_PASSWORD}@haproxy:${HAPROXY_PORT:-5984}/medic
      - BUILDS_URL=${MARKET_URL_READ:-https://staging.dev.medicmobile.org}/${BUILDS_SERVER:-_couch/builds}
      - UPGRADE_SERVICE_URL=${UPGRADE_SERVICE_URL:-http://localhost:5100}
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:

  cht-sentinel:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-sentinel:4.0.0-alpha
    depends_on:
      - haproxy
    environment:
      - COUCH_URL=http://${COUCHDB_USER:-admin}:${COUCHDB_PASSWORD}@haproxy:${HAPROXY_PORT:-5984}/medic
      - API_HOST=cht-api
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:

  cht-nginx:
    image: 720541322708.dkr.ecr.eu-west-2.amazonaws.com/medic/cht-nginx:4.0.0-alpha
    depends_on:
      - cht-api
      - haproxy
    ports:
      - "${NGINX_HTTP_PORT:-80}:80"
      - "${NGINX_HTTPS_PORT:-443}:443"
    volumes:
      - cht-ssl:/root/.acme.sh/
    environment:
      - "CERTIFICATE_MODE=${CERTIFICATE_MODE:-SELF_SIGNED}"
      - "SSL_CERT_FILE_PATH=${SSL_CERT_FILE_PATH:-/etc/nginx/private/cert.pem}"
      - "SSL_KEY_FILE_PATH=${SSL_KEY_FILE_PATH:-/etc/nginx/private/key.pem}"
      - "COMMON_NAME=${COMMON_NAME:-test-nginx.dev.medicmobile.org}"
      - "EMAIL=${EMAIL:-domains@medic.org}"
      - "COUNTRY=${COUNTRY:-US}"
      - "STATE=${STATE:-California}"
      - "LOCALITY=${LOCALITY:-San_Francisco}"
      - "ORGANISATION=${ORGANISATION:-medic}"
      - "DEPARTMENT=${DEPARTMENT:-Information_Security}"
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      cht-net:

networks:
  cht-net:

volumes:
  cht-ssl:
```

</details> 

and then run this command: `COMPOSE_PROJECT_NAME=testee COUCHDB_PASSWORD=password COUCHDB_USER=medic docker-compose -f docker-compose_cht-core.yml -f docker-comose_cht-couchdb.yml up -d` it should just work!

**Comment by mrjones-plip** [comment]
> isolated from the host

I forgot to speak to this!  It will be just as isolated (or just as not isolated) as if we didn't use the `COMPOSE_PROJECT_NAME` value.  That is, if we `expose:` no ports, then the host won't able to get in (aside from the fact that it could launch an container on that same network - but I assume that's out of scope)

**Comment by dianabarsan** [comment]
Thanks @mrjones-plip 
I already went through the custom network yesterday but didn't have time to also add an update here. This is the pull request from cht-upgrade-service: https://github.com/medic/cht-upgrade-service/pull/17
I want to create a branch in cht-core where we run the upgrade e2e test against this version of cht-upgrade-service, and it it works, I'll send this to the presses :)

**Comment by dianabarsan** [comment]
>  if you start the upgrade service with the same COMPOSE_PROJECT_NAME

I'm reluctant to do this, especially because we want to treat the upgrade service and the CHT containers separately: started separately, upgraded separately. The docker compose files for cht-core and cht-upgrade-service _need to be_ in different folders on disk. If someone wants to stop the cht-upgrade-service alone, and use docker-compose, they'd end up taking cht-core down as well. 

The alternative is to have two additional environment variables that we pass to the upgrade-service when starting: 
- CHT_COMPOSE_PROJECT - project name to be used when starting CHT containers 
- CHT_NETWORK_NAME - the network that the cht-upgrade-service and the cht-core containers would share 

Appreciate the feedback :)

**Comment by mrjones-plip** [comment]
> I'm reluctant to do this, especially because we want to treat the upgrade service and the CHT containers separately: started separately, upgraded separately

I might be miss-understanding the desired use case, but you can start the upgrade service in different directories with different `COMPOSE_PROJECT_NAME` values.  Here's a video illustrating the name space protections:

https://user-images.githubusercontent.com/8253488/194377607-1a770d7e-627c-481a-a6b5-60a6790733b3.mp4

To me this removes the need for your proposed `CHT_COMPOSE_PROJECT` and `CHT_NETWORK_NAME` env vars

**Comment by dianabarsan** [comment]
> with different COMPOSE_PROJECT_NAME values.

Your test proves the point exactly, where the "upgrader_upgrader_1" can't reach nginx over docker network. 
Only the upgrader containers that you started with the cht1 and cht2 project names were able to reach their corresponding nginx containers. 

What happens when you do:
`COMPOSE_PROJECT_NAME=cht1 docker-compose  down` in the upgrader folder?

**Comment by dianabarsan** [comment]
I'm trying the same experiment with docker-compose v2, and it's taking down both containers :( 
Could you please give that one a try as well?

**Comment by mrjones-plip** [comment]
Ah hah! good catch! I do see that compose v2 is more aggressive when shutting down other containers in the project now. 

I ...think this is still ok?  That is, I think we want to have an instance of cht-upgrade running per instance of core, right?

**Comment by dianabarsan** [comment]
yes, but I don't think we want to not be able to take them down individually.

**Comment by dianabarsan** [comment]
I'm leaning towards not using the same project name, and keeping cht-upgrade-service separate.

**Comment by mrjones-plip** [comment]
I haven't played enough with the upgrader docker stuffs to have an opinion on this just yet.  I trust your judgement on how to best proceed.

**Comment by dianabarsan** [comment]
Ok, cool, thanks for all the support. 
I have two PRs ready that add the `CHT_COMPOSE_PROJECT` and `CHT_NETWORK_NAME`, in both `cht-upgrade-service` and `cht`.

**Comment by dianabarsan** [comment]
I've merged this directly, because the update required changes in both `cht` and `cht-upgrade-service`, and once the upgrade-service was merged, the upgrade e2e in `cht` stopped working. 

AT should cover that: 
- the upgrade service starts and upgrades containers with default project name and default network
- the upgrade service starts and upgrades containers with custom project name and custom network
- users can start multiple CHT instances on the same machine
- users can start multiple CHT + upgrade service on the same machine (once https://github.com/medic/cht-upgrade-service/pull/18 lands)
